File: impacted-test-engine/src/main/java/com/teamscale/test_impacted/engine/ImpactedTestEngine.java
package com.teamscale.test_impacted.engine;

import com.teamscale.test_impacted.commons.LoggerUtils;
import com.teamscale.test_impacted.engine.options.TestEngineOptionUtils;
import com.teamscale.test_impacted.engine.options.TestEngineOptions;
import org.junit.platform.engine.EngineDiscoveryRequest;
import org.junit.platform.engine.ExecutionRequest;
import org.junit.platform.engine.TestDescriptor;
import org.junit.platform.engine.TestEngine;
import org.junit.platform.engine.UniqueId;

import java.util.Optional;
import java.util.logging.Logger;

/** Test engine for executing impacted tests. */
public class ImpactedTestEngine implements TestEngine {

	/** The id of the {@link ImpactedTestEngine}. */
	public static final String ENGINE_ID = "teamscale-test-impacted";

	public static final Logger LOGGER = LoggerUtils.getLogger(ImpactedTestEngine.class);

	private InternalImpactedTestEngine internalImpactedTestEngine = null;

	@Override
	public String getId() {
		return ENGINE_ID;
	}

	@Override
	public TestDescriptor discover(EngineDiscoveryRequest discoveryRequest, UniqueId uniqueId) {
		TestEngineOptions engineOptions = TestEngineOptionUtils
				.getEngineOptions(discoveryRequest.getConfigurationParameters());
		ImpactedTestEngineConfiguration configuration = engineOptions.createTestEngineConfiguration();

		// Re-initialize the configuration for this discovery (and optional following execution).
		internalImpactedTestEngine =
				new InternalImpactedTestEngine(configuration, engineOptions.getPartition());

		return internalImpactedTestEngine.discover(discoveryRequest, uniqueId);
	}

	@Override
	public void execute(ExecutionRequest request) {
		// According to the TestEngine interface the request must correspond to the last execution request. Therefore, we
		// may re-use the configuration initialized during discovery.
		if (internalImpactedTestEngine == null) {
			throw new AssertionError("Can't execute request without discovering it first.");
		}
		internalImpactedTestEngine.execute(request);
	}

	@Override
	public Optional<String> getGroupId() {
		return Optional.of("com.teamscale");
	}

	@Override
	public Optional<String> getArtifactId() {
		return Optional.of("impacted-test-engine");
	}
}

File: impacted-test-engine/src/main/kotlin/com/teamscale/test_impacted/engine/ImpactedTestEngine.kt
package com.teamscale.test_impacted.engine

import com.teamscale.test_impacted.commons.LoggerUtils.createLogger
import com.teamscale.test_impacted.engine.options.TestEngineOptionUtils
import org.junit.platform.engine.*
import org.junit.platform.engine.support.descriptor.EngineDescriptor
import java.util.*

/** Test engine for executing impacted tests.  */
class ImpactedTestEngine : TestEngine {
	private var internalImpactedTestEngine: InternalImpactedTestEngine? = null

	override fun getId() = ENGINE_ID

	override fun discover(
		discoveryRequest: EngineDiscoveryRequest,
		uniqueId: UniqueId
	): TestDescriptor {
		val engineOptions = TestEngineOptionUtils
			.getEngineOptions(discoveryRequest.configurationParameters)
		if (!engineOptions.enabled) {
			return EngineDescriptor(uniqueId, ENGINE_NAME)
		}
		val configuration = engineOptions.testEngineConfiguration
		val engine = InternalImpactedTestEngine(configuration, engineOptions.partition)

		// Re-initialize the configuration for this discovery (and optional following execution).
		internalImpactedTestEngine = engine

		return engine.discover(discoveryRequest, uniqueId)
	}

	override fun execute(request: ExecutionRequest) {
		// According to the TestEngine interface the request must correspond to the last execution request. Therefore, we
		// may re-use the configuration initialized during discovery.
		// Engine might be null if it is not enabled.
		internalImpactedTestEngine?.execute(request)
	}

	override fun getGroupId() = Optional.of("com.teamscale")

	override fun getArtifactId() = Optional.of("impacted-test-engine")

	companion object {
		/** The id of the [ImpactedTestEngine].  */
		const val ENGINE_ID = "teamscale-test-impacted"

		/** The human readable name of the [ImpactedTestEngine].  */
		const val ENGINE_NAME = "Teamscale Impacted Tests"

		/** Logger instance */
		val LOG = createLogger()
	}
}
